core/src/main/scala/org/apache/spark/sql/delta/GeneratedColumn.scala

```diff
@@ -296,27 +296,9 @@ object GeneratedColumn extends DeltaLogging with AnalysisHelper {
     val nameNormalizer: String => String =
       if (spark.sessionState.conf.caseSensitiveAnalysis) x => x else _.toLowerCase(Locale.ROOT)
 
-    /**
-     * If the column `a`'s type matches the expected type, call `func` to create
-     * `OptimizablePartitionExpression`. Returns a normalized column name with its
-     * `OptimizablePartitionExpression`
-     */
-    def checkTypeAndCreateExpr(
-        a: AttributeReference,
-        expectedType: DataType)(
-        func: => OptimizablePartitionExpression):
-      Option[(String, OptimizablePartitionExpression)] = {
-      // Technically, we should only refer to a column in the table schema. Check the column name
-      // here just for safety.
-      if (a.dataType == expectedType &&
-          schema.exists(f => nameEquality(f.name, a.name) && f.dataType == expectedType)) {
```

Comment:
I moved the data type check to the pattern matching, I wasn't sure what checking the schema added, aren't we operating off an analyzed projection from the schema for these checks?

Reply:
This is for safety. I don't expect that could happen. But it's possible, such as Spark changes an expression type from IntegerType to LongType, and this may cause our optimization no longer correct. Could you add this check back?

Reply:
@Kimahriman I finally recalled what we added this name check. We try to avoid referring to a column that doesn't exist in a table, such as https://github.com/apache/spark/blob/master/sql/catalyst/src/main/java/org/apache/spark/sql/connector/catalog/MetadataColumn.java#L40 , although we are using `LocalRelation` and that's impossible (but maybe something could be changed in future). Maybe just add the check back for safety?

Reply:
Added that back in, had to maintain the name parts as a sequence for longer and just used the builtin `findNestedField`